src/util/util.c

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2018 King's College London
+// created by the Software Development Team <http://soft-dev.org/>
+//
+// The Universal Permissive License (UPL), Version 1.0
+//
+// Subject to the condition set forth below, permission is hereby granted to any
+// person obtaining a copy of this software, associated documentation and/or
+// data (collectively the "Software"), free of charge and under any and all
+// copyright rights in the Software, and any and all patent rights owned or
+// freely licensable by each licensor hereunder covering either (i) the
+// unmodified Software as contributed to or provided by such licensor, or (ii)
+// the Larger Works (as defined below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file
+// if one is included with the Software (each a "Larger Work" to which the Software
+// is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#define _GNU_SOURCE
+
+#include <link.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <libgen.h>
+
+// Private prototypes.
+static int dl_iterate_phdr_cb(struct dl_phdr_info *, size_t, void *);
+
+// Exposed prototypes.
+int traceme_exec_base(uintptr_t *);
+
+/*
+ * Called by `dl_iterate_phdr` once for each program header.
+ *
+ * We are looking for the start address of the program's executable
+ * instructions.
+ */
+static int
+dl_iterate_phdr_cb(struct dl_phdr_info *info, size_t size, void *data)
+{
+#ifdef __OpenBSD__
+    Elf_Phdr phdr;
+    Elf_Half i;
+#else
+    ElfW(Phdr) phdr;
+    ElfW(Half) i;
+#endif
+    (void) size;
+
+    // Check if this is the entry for the binary itself (not a shared object).
+#if defined(__linux__)
+    // On Linux the main binary has no name.
+    if (info->dlpi_name[0] != '\0') {
+        return 0;
```

Comment:
Hang on, `0` is the success return code I assume, but it seems to being returned here on failure?

Reply:
This is the interface required by the `dl_iterate_phdr` interface.

```
      The dl_iterate_phdr() function walks through the list  of  an  applica‐
       tion's  shared  objects  and  calls the function callback once for each
       object, until either all shared objects have been processed or callback
       returns a nonzero value.
```

We can invert the logic in `traceme_exec_base` if you like though.

Shall I move the explanation?

Reply:
Maybe move `traceme_exec_base` above this function (assuming that most people read from top to bottom!)?

Reply:
Sure thing!